src/Raven.Server/Documents/AttachmentsStorage.Sharding.cs

```diff
@@ -0,0 +1,71 @@
+﻿using Raven.Server.Documents.Replication.ReplicationItems;
+using System.Collections.Generic;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Binary;
+using Sparrow.Server;
+using Sparrow.Server.Utils;
+using Voron;
+using Voron.Data.Tables;
+using static Raven.Server.Documents.Schemas.Attachments;
+
+namespace Raven.Server.Documents
+{
+    public unsafe partial class AttachmentsStorage
+    {
+        public IEnumerable<ReplicationBatchItem> GetAttachmentsByBucketFrom(DocumentsOperationContext context, int bucket, long etag)
+        {
+            var table = context.Transaction.InnerTransaction.OpenTable(AttachmentsSchema, AttachmentsMetadataSlice);
+
+            foreach (var result in ShardedDocumentsStorage.GetItemsByBucket(context.Allocator, table, AttachmentsSchema.DynamicKeyIndexes[AttachmentsBucketAndEtagSlice], bucket, etag))
+            {
+                var attachment = TableValueToAttachment(context, ref result.Result.Reader);
+
+                var stream = GetAttachmentStream(context, attachment.Base64Hash);
+                if (stream == null)
+                    ThrowMissingAttachment(attachment.Name);
+
+                attachment.Stream = stream;
+
+                yield return AttachmentReplicationItem.From(context, attachment);
+            }
+        }
+
+        [StorageIndexEntryKeyGenerator]
+        internal static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForAttachments(ByteStringContext context, ref TableValueReader tvr, out Slice slice)
+        {
+            return ShardedDocumentsStorage.ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, (int)AttachmentsTable.LowerDocumentIdAndLowerNameAndTypeAndHashAndContentType,
+                (int)AttachmentsTable.Etag, ref tvr, out slice);
+        }
+
+        private void UpdateBucketStatsOnPutOrDeleteStream(DocumentsOperationContext context, Slice attachmentKey, long sizeChange)
+        {
+            if (_documentDatabase is not ShardedDocumentDatabase)
```

Comment:
That is a bit tricky I think. Can we throw exception here?

Reply:
we reach this code on every put/delete stream, from non-sharded document databases as well..
I'll try to change that